src/modules/iotjs_module_buffer.c

```diff
@@ -88,31 +93,122 @@ static size_t bound_range(size_t index, size_t low, size_t upper) {
 }
 
 
-static int8_t hex2bin(char c) {
-  if (c >= '0' && c <= '9')
+static int8_t hexToBin(char c) {
+  if (c >= '0' && c <= '9') {
     return (int8_t)(c - '0');
-  if (c >= 'A' && c <= 'F')
+  }
+  if (c >= 'A' && c <= 'F') {
     return (int8_t)(10 + (c - 'A'));
-  if (c >= 'a' && c <= 'f')
+  }
+  if (c >= 'a' && c <= 'f') {
     return (int8_t)(10 + (c - 'a'));
+  }
 
   return (int8_t)(-1);
 }
 
 
 static size_t hex_decode(char* buf, size_t len, const char* src,
                          const size_t srcLen) {
-  size_t i;
-
-  for (i = 0; i < len && i * 2 + 1 < srcLen; ++i) {
-    int8_t a = hex2bin(src[i * 2 + 0]);
-    int8_t b = hex2bin(src[i * 2 + 1]);
-    if (a == -1 || b == -1)
-      return i;
-    buf[i] = (a << 4) | b;
+  const char* bufStart = buf;
+  const char* bufEnd = buf + len;
+  const char* srcEnd = src + srcLen;
+
+  if ((srcLen & 0x1) != 0) {
+    return 0;
+  }
+
+  while (src < srcEnd) {
+    int8_t a = hexToBin(src[0]);
+    int8_t b = hexToBin(src[1]);
+
+    if (a == -1 || b == -1) {
+      return 0;
+    }
+
+    if (buf < bufEnd) {
+      *buf++ = (a << 4) | b;
+    }
+
+    src += 2;
+  }
+
+  return (size_t)((buf - bufStart) + 1);
+}
+
+
+static int32_t base64ToBin(char c) {
+  if (c >= 'A' && c <= 'Z') {
+    return (int32_t)(c - 'A');
+  }
+  if (c >= 'a' && c <= 'z') {
+    return (int32_t)(26 + (c - 'a'));
+  }
+  if (c >= '0' && c <= '9') {
+    return (int32_t)(52 + (c - '0'));
+  }
+  if (c == '+') {
+    return 62;
+  }
+  if (c == '/') {
+    return 63;
+  }
+
+  return (int32_t)(-1);
+}
+
+
+static size_t base64_decode(char* buf, size_t len, const char* src,
+                            const size_t srcLen) {
+  if (srcLen == 0) {
+    return 0 + 1;
```

Comment:
It seems weird that the size decoded string is 1 when srcLen is 0. This causes that we need to reduce the returned value like `jerry_create_number(nbytes - 1);`.

Reply:
The function returns with 0 on error, and len + 1 otherwise. I could use `ssize_t` and use -1 on error, but loosing half of the range of `size_t` for returning an error seems a waste. What do you prefer?

Reply:
The current one is now fine with me. Thanks.